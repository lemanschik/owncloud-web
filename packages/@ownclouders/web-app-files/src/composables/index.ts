export * from './actions'
export * from './archiverService'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './shares'
export * from './spaces'
export * from './viewMode'
