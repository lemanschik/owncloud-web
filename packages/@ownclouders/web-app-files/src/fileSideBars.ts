import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import SharesPanel from './components/SideBar/Shares/SharesPanel.vue'
import TagsPanel from './components/SideBar/TagsPanel.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import SpaceDetails from 'web-pkg/src/components/sideBar/Spaces/Details/SpaceDetails.vue'
import SpaceNoSelection from 'web-pkg/src/components/sideBar/Spaces/SpaceNoSelection.vue'
import {
  isLocationTrashActive,
  isLocationPublicActive,
  isLocationSpacesActive,
  isLocationSharesActive
} from './router'
import { spaceRoleEditor, spaceRoleManager } from 'web-client/src/helpers/share'

import { Panel } from '../../web-pkg/src/components/sideBar'

import { Resource, User } from 'web-client'
import { Router } from 'vue-router'
import SpaceDetailsMultiple from 'web-pkg/src/components/sideBar/Spaces/Details/SpaceDetailsMultiple.vue'

function $gettext(msg: string): string {
  return msg
}

const panelGenerators: (({
  rootFolder,
  resource,
  selectedFiles,
  router,
  multipleSelection,
  user,
  capabilities
}: {
  rootFolder: boolean
  resource: Resource
  selectedFiles: Resource[]
  router: Router
  multipleSelection: boolean
  user: User
  capabilities: any
}) => Panel)[] = [
  // We don't have file details in the trashbin, yet.
  // Only allow `actions` panel on trashbin route for now.
  ({ router, multipleSelection, rootFolder, resource }): Panel => ({
    app: 'no-selection',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: NoSelection,
    default: () => true,
    get enabled() {
      return (
        !multipleSelection &&
        !isLocationSpacesActive(router, 'files-spaces-projects') &&
        (!resource || (rootFolder && resource?.type !== 'space'))
      )
    }
  }),
  ({ router, resource, multipleSelection }): Panel => ({
    app: 'no-selection',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceNoSelection,
    default: () => true,
    get enabled() {
      return (
        isLocationSpacesActive(router, 'files-spaces-projects') && !multipleSelection && !resource
      )
    }
  }),
  ({ router, multipleSelection, rootFolder, resource }) => ({
    app: 'details',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetails,
    default: !isLocationTrashActive(router, 'files-trash-generic'),
    get enabled() {
      return !!(
        !isLocationTrashActive(router, 'files-trash-generic') &&
        !multipleSelection &&
        !rootFolder &&
        resource &&
        resource.type !== 'space'
      )
    }
  }),
  ({ multipleSelection, router }) => ({
    app: 'details-multiple',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetailsMultiple,
    componentAttrs: {
      get showSpaceCount() {
        return (
          !isLocationSpacesActive(router, 'files-spaces-generic') &&
          !isLocationSharesActive(router, 'files-shares-with-me') &&
          !isLocationTrashActive(router, 'files-trash-generic')
        )
      }
    },
    default: () => true,
    get enabled() {
      return multipleSelection && !isLocationSpacesActive(router, 'files-spaces-projects')
    }
  }),
  ({ multipleSelection, selectedFiles, router }) => ({
    app: 'details-space-multiple',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetailsMultiple,
    componentAttrs: {
      selectedSpaces: selectedFiles
    },
    default: () => true,
    get enabled() {
      return multipleSelection && isLocationSpacesActive(router, 'files-spaces-projects')
    }
  }),
  ({ multipleSelection, resource }) => ({
    app: 'details-space',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetails,
    default: () => true,
    get enabled() {
      return resource?.type === 'space' && !multipleSelection
    }
  }),
  ({ router, multipleSelection, rootFolder, resource }) => ({
    app: 'actions',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: FileActions,
    default: isLocationTrashActive(router, 'files-trash-generic'),
    get enabled() {
      return !!(!multipleSelection && !rootFolder && resource && resource.type !== 'space')
    }
  }),
  ({ multipleSelection, resource, user }) => ({
    app: 'space-actions',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: SpaceActions,
    get enabled() {
      if (multipleSelection) {
        return false
      }
      if (resource?.type !== 'space') {
        return false
      }
      return !![
        ...resource.spaceRoles[spaceRoleManager.name],
        ...resource.spaceRoles[spaceRoleEditor.name]
      ].find((role) => role.id === user.uuid)
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder, resource }) => ({
    app: 'sharing',
    icon: 'user-add',
    iconFillType: 'line',
    title: $gettext('Shares'),
    component: SharesPanel,
    componentAttrs: {
      showSpaceMembers: false,
      get showLinks() {
        if (capabilities.files_sharing) {
          return capabilities.files_sharing.public.enabled
        }
        return false
      }
    },
    get enabled() {
      if (multipleSelection || rootFolder || !resource || resource.type === 'space') {
        return false
      }
      if (
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.api_enabled
      }
      return false
    }
  }),
  ({ capabilities, resource, router, multipleSelection, rootFolder }) => ({
    app: 'tags',
    icon: 'price-tag-3',
    iconFillType: 'line',
    title: $gettext('Tags'),
    component: TagsPanel,
    componentAttrs: {},
    get enabled() {
      if (
        !capabilities?.files?.tags ||
        multipleSelection ||
        rootFolder ||
        !resource ||
        resource.type === 'space'
      ) {
        return false
      }
      if (typeof resource.canEditTags !== 'function' || !resource.canEditTags()) {
        return false
      }
      return !(
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      )
    }
  }),
  ({ multipleSelection, resource, capabilities }) => ({
    app: 'space-share',
    icon: 'group',
    title: $gettext('Members'),
    component: SharesPanel,
    componentAttrs: {
      showSpaceMembers: true,
      get showLinks() {
        if (capabilities.files_sharing) {
          return capabilities.files_sharing.public.enabled
        }
        return false
      }
    },
    get enabled() {
      return resource?.type === 'space' && !multipleSelection
    }
  }),
  ({ capabilities, resource, router, multipleSelection, rootFolder }) => ({
    app: 'versions',
    icon: 'git-branch',
    title: $gettext('Versions'),
    component: FileVersions,
    get enabled() {
      if (multipleSelection || rootFolder) {
        return false
      }
      if (
        isLocationTrashActive(router, 'files-trash-generic') ||
        isLocationPublicActive(router, 'files-public-link')
      ) {
        return false
      }
      return (
        !!capabilities.core && resource && resource.type !== 'folder' && resource.type !== 'space'
      )
    }
  })
]

export default panelGenerators
