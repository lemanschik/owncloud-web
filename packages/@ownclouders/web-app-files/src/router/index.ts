import { RouteLocationNamedRaw } from 'vue-router'

import {
  buildRoutes as buildCommonRoutes,
  isLocationCommonActive,
  createLocationCommon
} from './common'
import { buildRoutes as buildDeprecatedRoutes, isLocationActive } from './deprecated'
import {
  buildRoutes as buildPublicRoutes,
  createLocationPublic,
  isLocationPublicActive
} from './public'
import { RouteComponents } from './router'
import {
  buildRoutes as buildSharesRoutes,
  isLocationSharesActive,
  createLocationShares
} from './shares'
import {
  buildRoutes as buildSpacesRoutes,
  isLocationSpacesActive,
  createLocationSpaces
} from './spaces'
import {
  buildRoutes as buildTrashRoutes,
  isLocationTrashActive,
  createLocationTrash
} from './trash'
import type { ActiveRouteDirectorFunc } from './utils'

const ROOT_ROUTE = {
  name: 'root',
  path: '/',
  redirect: (to) => createLocationSpaces('files-spaces-generic', to)
}

const buildRoutes = (components: RouteComponents): RouteLocationNamedRaw[] => [
  ROOT_ROUTE,
  ...buildCommonRoutes(components),
  ...buildSharesRoutes(components),
  ...buildPublicRoutes(components),
  ...buildSpacesRoutes(components),
  ...buildTrashRoutes(components),
  ...buildDeprecatedRoutes()
]

export {
  createLocationCommon,
  createLocationShares,
  createLocationSpaces,
  createLocationPublic,
  isLocationCommonActive,
  isLocationSharesActive,
  isLocationSpacesActive,
  isLocationPublicActive,
  isLocationActive,
  isLocationTrashActive,
  createLocationTrash,
  buildRoutes,
  ActiveRouteDirectorFunc
}
