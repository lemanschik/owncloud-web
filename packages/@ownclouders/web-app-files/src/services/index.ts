export * from './archiver'
export { default as Registry } from './registry'
