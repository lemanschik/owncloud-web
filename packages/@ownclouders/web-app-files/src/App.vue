<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="files-wrapper oc-width-expand oc-height-1-1 oc-flex-wrap" />
  </main>
</template>
<script lang="ts">
import { defineComponent, onBeforeUnmount, watch, ref } from 'vue'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    const dragareaEnabled = ref(false)
    watch(useRoute(), () => {
      store.dispatch('Files/resetFileSelection')
    })

    const hideDropzone = () => {
      dragareaEnabled.value = false
    }
    const onDragOver = (event) => {
      dragareaEnabled.value = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }

    const dragOver = eventBus.subscribe('drag-over', onDragOver)
    const dragOut = eventBus.subscribe('drag-out', hideDropzone)
    const drop = eventBus.subscribe('drop', hideDropzone)

    onBeforeUnmount(() => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
    return { dragareaEnabled }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
