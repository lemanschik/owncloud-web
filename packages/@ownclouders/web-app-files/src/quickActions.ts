import { createQuicklink } from './helpers/share'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { SideBarEventTopics } from 'web-pkg/src/composables/sideBar'
import { Resource } from 'web-client'
import { Language } from 'vue3-gettext'
import { Ability, ClientService } from 'web-pkg'
import { Store } from 'vuex'

export function canShare(item, store) {
  const { capabilities } = store.state.user
  if (!capabilities.files_sharing || !capabilities.files_sharing.api_enabled) {
    return false
  }
  if (item.isReceivedShare() && !capabilities.files_sharing.resharing) {
    return false
  }
  return item.canShare()
}

export function showQuickLinkPasswordModal({ $gettext, store }, onConfirm) {
  const modal = {
    variation: 'passive',
    title: $gettext('Set password'),
    cancelText: $gettext('Cancel'),
    confirmText: $gettext('Set'),
    hasInput: true,
    inputDescription: $gettext('Passwords for links are required.'),
    inputLabel: $gettext('Password'),
    inputType: 'password',
    onCancel: () => store.dispatch('hideModal'),
    onConfirm: async (password) => {
      if (!password || password.trim() === '') {
        store.dispatch('showMessage', {
          title: $gettext('Password cannot be empty'),
          status: 'danger'
        })
      } else {
        await store.dispatch('hideModal')
        onConfirm(password)
      }
    },
    onInput: (password) => {
      if (password.trim() === '') {
        return store.dispatch('setModalInputErrorMessage', $gettext('Password cannot be empty'))
      }
      return store.dispatch('setModalInputErrorMessage', null)
    }
  }

  return store.dispatch('createModal', modal)
}

interface QuickLinkContext {
  ability: Ability
  clientService: ClientService
  item: Resource
  language: Language
  store: Store<any>
}

export default {
  collaborators: {
    id: 'collaborators',
    label: ($gettext) => $gettext('Add people'),
    icon: 'user-add',
    iconFillType: undefined,
    handler: () => eventBus.publish(SideBarEventTopics.openWithPanel, 'sharing#peopleShares'),
    displayed: canShare
  },
  quicklink: {
    id: 'quicklink',
    label: ($gettext) => $gettext('Copy link'),
    icon: 'link',
    iconFillType: undefined,
    handler: async ({ ability, clientService, item, language, store }: QuickLinkContext) => {
      const passwordEnforced =
        store.getters.capabilities?.files_sharing?.public?.password?.enforced_for?.read_only ===
        true

      if (passwordEnforced) {
        return showQuickLinkPasswordModal(
          { store, $gettext: language.$gettext },
          async (password) => {
            await createQuicklink({
              ability,
              clientService,
              language,
              password,
              resource: item,
              store
            })
          }
        )
      }

      await createQuicklink({
        ability,
        clientService,
        language,
        resource: item,
        store
      })
    },
    displayed: canShare
  }
}
