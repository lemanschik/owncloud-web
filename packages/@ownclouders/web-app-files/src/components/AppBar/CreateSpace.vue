<template>
  <oc-button
    id="new-space-menu-btn"
    key="new-space-menu-btn-enabled"
    v-oc-tooltip="$gettext('Create a new space')"
    :aria-label="$gettext('Create a new space')"
    appearance="filled"
    variation="primary"
    @click="showCreateSpaceModal"
  >
    <oc-icon name="add" />
    <span v-translate>New Space</span>
  </oc-button>
</template>

<script lang="ts">
import { mapActions, mapMutations } from 'vuex'
import { defineComponent } from 'vue'
import { useClientService } from 'web-pkg/src/composables'
import { useCreateSpace } from 'web-app-files/src/composables'
import { useSpaceHelpers } from 'web-pkg/src/composables/spaces'

export default defineComponent({
  setup() {
    const clientService = useClientService()
    const { createSpace } = useCreateSpace()
    const { checkSpaceNameModalInput } = useSpaceHelpers()
    return { clientService, createSpace, checkSpaceNameModalInput }
  },
  methods: {
    ...mapActions(['showMessage', 'createModal', 'hideModal', 'setModalInputErrorMessage']),
    ...mapMutations('runtime/spaces', ['UPSERT_SPACE']),
    ...mapMutations('Files', ['UPSERT_RESOURCE', 'UPDATE_RESOURCE_FIELD']),

    showCreateSpaceModal() {
      const modal = {
        variation: 'passive',
        title: this.$gettext('Create a new space'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.$gettext('Create'),
        hasInput: true,
        inputLabel: this.$gettext('Space name'),
        inputValue: this.$gettext('New space'),
        onCancel: this.hideModal,
        onConfirm: this.addNewSpace,
        onInput: this.checkSpaceNameModalInput
      }

      this.createModal(modal)
    },

    async addNewSpace(name) {
      try {
        const createdSpace = await this.createSpace(name)
        this.hideModal()
        this.UPSERT_RESOURCE(createdSpace)
        this.UPSERT_SPACE(createdSpace)
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettext('Creating space failedâ€¦'),
          status: 'danger'
        })
      }
    }
  }
})
</script>
