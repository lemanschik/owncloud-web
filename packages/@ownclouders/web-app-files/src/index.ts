import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { SDKSearch } from './search'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { Registry, ArchiverService } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'
import { AppNavigationItem, AppReadyHookArgs } from 'web-pkg/src/apps'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import { isPersonalSpaceResource } from 'web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find((drive) =>
          isPersonalSpaceResource(drive)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/share` }],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0'
      }
    }
  ]
}

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    FilesDrop,
    SearchResults,
    Shares: {
      SharedViaLink,
      SharedWithMe,
      SharedWithOthers
    },
    Spaces: {
      DriveResolver: SpaceDriveResolver,
      Projects: SpaceProjects
    },
    Trash: {
      Overview: TrashOverview
    }
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store, globalProperties }: AppReadyHookArgs) {
    const { $clientService } = globalProperties
    Registry.sdkSearch = new SDKSearch(store, router, $clientService)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    eventBus.publish('app.search.register.provider', Registry.sdkSearch)
    globalProperties.$archiverService = new ArchiverService(
      $clientService,
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
  }
}
