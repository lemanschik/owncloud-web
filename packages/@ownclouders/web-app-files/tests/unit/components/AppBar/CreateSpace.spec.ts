import CreateSpace from '../../../../src/components/AppBar/CreateSpace.vue'
import { mockDeep } from 'jest-mock-extended'
import { Resource } from 'web-client'
import { Drive } from 'web-client/src/generated'
import {
  createStore,
  defaultPlugins,
  mount,
  defaultStoreMockOptions,
  defaultComponentMocks
} from 'web-test-helpers'

const selectors = {
  newSpaceBtn: '#new-space-menu-btn'
}

describe('CreateSpace component', () => {
  it('should show the "New Space" button', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.find(selectors.newSpaceBtn).exists()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should show a modal when clicking the "New Space" button', async () => {
    const { wrapper, storeOptions } = getWrapper()
    await wrapper.find(selectors.newSpaceBtn).trigger('click')
    expect(storeOptions.actions.createModal).toHaveBeenCalledTimes(1)
  })
  describe('method "addNewSpace"', () => {
    it('creates the space and updates the readme data after creation', async () => {
      const { wrapper, mocks, storeOptions } = getWrapper()
      const graphMock = mocks.$clientService.graphAuthenticated
      const drive = mockDeep<Drive>()
      graphMock.drives.createDrive.mockResolvedValue(drive as any)
      graphMock.drives.updateDrive.mockResolvedValue(drive as any)
      mocks.$clientService.webdav.putFileContents.mockResolvedValue(mockDeep<Resource>())
      await wrapper.vm.addNewSpace('New space')
      expect(storeOptions.modules.runtime.modules.spaces.mutations.UPSERT_SPACE).toHaveBeenCalled()
    })
    it('shows a message when an error occurred', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper, mocks, storeOptions } = getWrapper()
      const graphMock = mocks.$clientService.graphAuthenticated
      graphMock.drives.createDrive.mockRejectedValue({})
      await wrapper.vm.addNewSpace('New space')
      expect(storeOptions.actions.showMessage).toHaveBeenCalledWith(
        expect.anything(),
        expect.objectContaining({
          status: 'danger',
          title: 'Creating space failedâ€¦'
        })
      )
    })
  })
})

function getWrapper() {
  const storeOptions = { ...defaultStoreMockOptions }
  const store = createStore(storeOptions)
  const mocks = defaultComponentMocks()
  return {
    storeOptions,
    mocks,
    wrapper: mount(CreateSpace, {
      global: {
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
