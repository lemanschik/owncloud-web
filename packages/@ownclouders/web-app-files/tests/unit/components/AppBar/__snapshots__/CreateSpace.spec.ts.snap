// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSpace component should show the "New Space" button 1`] = `
<button aria-label="Create a new space" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled" id="new-space-menu-btn" type="button">
  <!-- @slot Content of the button -->
  <span class="oc-icon oc-icon-m oc-icon-passive">
    <!---->
  </span>
  <span data-current-language="en" data-msgid="New Space">New Space</span>
</button>
`;
